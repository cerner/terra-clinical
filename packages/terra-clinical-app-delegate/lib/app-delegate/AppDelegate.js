'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _DisclosureComponentRegistry = require('../disclosure-component-registry/DisclosureComponentRegistry');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var supportedAttributes = ['disclose', 'dismiss', 'closeDisclosure', 'goBack', 'maximize', 'minimize'];

var AppDelegateInstance = function AppDelegateInstance(data) {
  var _this = this;

  _classCallCheck(this, AppDelegateInstance);

  supportedAttributes.forEach(function (attribute) {
    if (data[attribute]) {
      _this[attribute] = data[attribute];
    }
  });
};

var create = function create(data) {
  return Object.freeze(new AppDelegateInstance(data));
};

var clone = function clone(delegate, data) {
  var ancestorDelegate = delegate || {};

  var mergedData = {};
  supportedAttributes.forEach(function (attribute) {
    mergedData[attribute] = data[attribute] || ancestorDelegate[attribute];
  });

  return create(mergedData);
};

var isEqual = function isEqual(delegateA, delegateB) {
  if (!delegateA || !delegateB) {
    return false;
  }

  if (delegateA === delegateB) {
    return true;
  }

  return !supportedAttributes.some(function (attribute) {
    return delegateA[attribute] !== delegateB[attribute];
  });
};

var AppDelegate = {
  propType: _propTypes2.default.instanceOf(AppDelegateInstance),
  create: create,
  clone: clone,
  isEqual: isEqual,
  registerComponentForDisclosure: _DisclosureComponentRegistry.registerComponentForDisclosure,
  getComponentForDisclosure: _DisclosureComponentRegistry.getComponentForDisclosure
};

exports.default = AppDelegate;