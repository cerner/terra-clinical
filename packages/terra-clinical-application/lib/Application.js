'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _terraBase = require('terra-base');

var _terraBase2 = _interopRequireDefault(_terraBase);

var _terraClinicalAppDelegate = require('terra-clinical-app-delegate');

var _terraClinicalAppDelegate2 = _interopRequireDefault(_terraClinicalAppDelegate);

require('terra-base/lib/baseStyles');

require('./Application.scss');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var propTypes = {
  /**
   * The AppDelegate instance that will be propagated to the Application's children.
   **/
  app: _terraClinicalAppDelegate2.default.propType,

  /**
   * The components to display within the Application.
   **/
  children: _propTypes2.default.node.isRequired
};

var Application = function Application(_ref) {
  var app = _ref.app,
      children = _ref.children,
      customProps = _objectWithoutProperties(_ref, ['app', 'children']);

  var childrenToRender = children;

  if (app) {
    childrenToRender = _react2.default.Children.map(children, function (child) {
      return _react2.default.cloneElement(child, { app: app });
    });
  }

  return _react2.default.createElement(
    _terraBase2.default,
    _extends({}, customProps, { className: (0, _classnames2.default)([customProps.className, 'terraClinical-Application']) }),
    childrenToRender
  );
};

Application.propTypes = propTypes;

exports.default = Application;