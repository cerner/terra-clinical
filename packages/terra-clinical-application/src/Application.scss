:local {
  .application {
    height: 100%;
  }
}
