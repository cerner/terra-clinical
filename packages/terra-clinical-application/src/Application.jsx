import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import Base from 'terra-base';
import AppDelegate from 'terra-app-delegate';
import styles from './Application.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * The AppDelegate instance that will be propagated to the Application's children.
   */
  app: AppDelegate.propType,

  /**
   * The components to display within the Application.
   */
  children: PropTypes.node.isRequired,
};

const Application = ({ app, children, ...customProps }) => {
  let childrenToRender = children;

  if (app) {
    childrenToRender = React.Children.map(children, child => (
      React.cloneElement(child, { app })
    ));
  }

  return (
    <Base {...customProps} className={cx(['application', customProps.className])}>
      {childrenToRender}
    </Base>
  );
};

Application.propTypes = propTypes;

export default Application;
