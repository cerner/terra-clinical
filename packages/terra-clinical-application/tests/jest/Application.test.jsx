import React from 'react';
import AppDelegate from 'terra-app-delegate';

import Application from '../../lib/Application';

it('should render default component', () => {
  const application = (
    <Application>
      <div>TestApp</div>
    </Application>
  );

  const wrapper = shallow(application);
  expect(wrapper).toMatchSnapshot();
});

it('should render custom properties on component', () => {
  const application = (
    <Application className="custom-class" id="my-id">
      <div>TestApp</div>
    </Application>
  );

  const wrapper = shallow(application);
  expect(wrapper).toMatchSnapshot();
});

it('should render children with app prop', () => {
  const appDelegate = AppDelegate.create({});

  const application = (
    <Application app={appDelegate}>
      <div>TestApp</div>
    </Application>
  );

  const wrapper = shallow(application);
  expect(wrapper).toMatchSnapshot();
});
