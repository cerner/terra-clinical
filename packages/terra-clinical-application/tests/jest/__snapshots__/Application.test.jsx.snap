// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render children with app prop 1`] = `
<Base
  className="application"
  customMessages={Object {}}
>
  <div
    app={AppDelegateInstance {}}
  >
    TestApp
  </div>
</Base>
`;

exports[`should render custom properties on component 1`] = `
<Base
  className="application custom-class"
  customMessages={Object {}}
  id="my-id"
>
  <div>
    TestApp
  </div>
</Base>
`;

exports[`should render default component 1`] = `
<Base
  className="application"
  customMessages={Object {}}
>
  <div>
    TestApp
  </div>
</Base>
`;
